File: VerbumKit/Sources/Clients/HelloAO/HelloAOBibleClient.swift
import Foundation
import Models

/// Reads Scripture from bible.helloao.org — free, keyless, open-licensed
/// translations served as static JSON — and keeps every chapter it has seen on
/// disk so it reads again offline (spec §39).
public struct HelloAOBibleClient: Sendable {
    /// Fetches the bytes at a URL. Injected so tests never touch the network.
    public typealias Transport = @Sendable (URL) async throws -> Data

    public let translationId: String
    let transport: Transport
    let cache: ChapterCache

    public init(translationId: String, transport: @escaping Transport = Self.urlSession, cache: ChapterCache = .onDisk) {
        self.translationId = translationId
        self.transport = transport
        self.cache = cache
    }

    public static let baseURL = URL(string: "https://bible.helloao.org/api")!

    /// Plain `URLSession`: a GET of a static file is all this API needs.
    public static let urlSession: Transport = { url in
        let (data, response) = try await URLSession.shared.data(from: url)
        if let http = response as? HTTPURLResponse, http.statusCode == 404 {
            throw BibleClientError.contentUnavailable(PassageReference(bookId: url.lastPathComponent, chapter: 0))
        }
        return data
    }

    public func chapter(bookId: BookID, chapter: Int) async throws -> [BiblePassage] {
        guard let book = BibleBook.book(id: bookId) else { throw BibleClientError.unknownBook(bookId) }
        let reference = PassageReference(bookId: bookId, chapter: chapter)
        guard (1...book.chapterCount).contains(chapter), let usfm = HelloAOBooks.usfmByOSIS[bookId] else {
            throw BibleClientError.contentUnavailable(reference)
        }
        if let cached = await cache.read(translationId: translationId, reference: reference),
           !cached.isEmpty, cached.allSatisfy({ $0.translationId == translationId }) {
            return cached
        }
        let url = Self.baseURL.appendingPathComponent("\(translationId)/\(usfm)/\(chapter).json")
        let data: Data
        do {
            data = try await transport(url)
        } catch let error as BibleClientError {
            throw error
        } catch {
            throw BibleClientError.networkUnavailable
        }
        let passages = try HelloAOChapter.passages(from: data, bookId: bookId)
        guard !passages.isEmpty, passages.allSatisfy({ $0.translationId == translationId }) else { throw BibleClientError.contentUnavailable(reference) }
        await cache.write(passages, translationId: translationId, reference: reference)
        return passages
    }

    public func passage(_ reference: PassageReference) async throws -> BiblePassage {
        let verses = try await chapter(bookId: reference.bookId, chapter: reference.chapter)
        let range = reference.verses ?? 1...verses.count
        guard range.upperBound <= verses.count else {
            throw BibleClientError.verseOutOfRange(reference, available: verses.count)
        }
        let selected = verses[(range.lowerBound - 1)..<range.upperBound]
        let rangeText = range.lowerBound == range.upperBound ? "\(range.lowerBound)" : "\(range.lowerBound)-\(range.upperBound)"
        return BiblePassage(
            id: "\(translationId):\(reference.bookId).\(reference.chapter).\(rangeText)",
            translationId: translationId,
            bookId: reference.bookId,
            chapter: reference.chapter,
            verseStart: range.lowerBound,
            verseEnd: range.upperBound,
            text: selected.map(\.text).joined(separator: " ")
        )
    }
}

/// Chapters already read, as JSON files in Caches. Small, simple, enough for
/// offline re-reading until a real local store (SQLite/SwiftData) is needed.
public actor ChapterCache {
    private let directory: URL?
    private var memory: [String: [BiblePassage]] = [:]
    private var recency: [String] = []
    private let memoryLimit = 24

    public init(directory: URL?) {
        self.directory = directory
    }

    public static let onDisk = ChapterCache(
        directory: FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.appendingPathComponent("Scripture", isDirectory: true)
    )
    /// A fresh, empty cache — one per test.
    public static var inMemory: ChapterCache { ChapterCache(directory: nil) }

    func read(translationId: String, reference: PassageReference) -> [BiblePassage]? {
        let key = key(translationId, reference)
        if let hit = memory[key] { remember(hit, for: key); return hit }
        guard let url = fileURL(key), let data = try? Data(contentsOf: url),
              let passages = try? JSONDecoder().decode([BiblePassage].self, from: data) else { return nil }
        remember(passages, for: key)
        return passages
    }

    func write(_ passages: [BiblePassage], translationId: String, reference: PassageReference) {
        let key = key(translationId, reference)
        remember(passages, for: key)
        guard let url = fileURL(key), let data = try? JSONEncoder().encode(passages) else { return }
        try? FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? data.write(to: url, options: .atomic)
    }

    private func key(_ translationId: String, _ reference: PassageReference) -> String {
        "\(translationId)_\(reference.bookId)_\(reference.chapter)"
    }

    private func remember(_ passages: [BiblePassage], for key: String) {
        recency.removeAll { $0 == key }
        recency.append(key)
        memory[key] = passages
        if recency.count > memoryLimit { memory.removeValue(forKey: recency.removeFirst()) }
    }

    private func fileURL(_ key: String) -> URL? {
        directory?.appendingPathComponent("\(key).json")
    }
}

extension BibleClient {
    /// Production client: helloao in the device language, with two safety
    /// nets — chapters read before (disk cache) and, for English, the bundled
    /// WEB — so the reader never dead-ends without a network.
    public static func live(language: BookLanguage, remote: HelloAOBibleClient? = nil) -> BibleClient {
        let remote = remote ?? HelloAOBibleClient(translationId: HelloAOTranslation.id(for: language))
        let offline: BibleClient? = language == .english ? .bundled : nil

        @Sendable func chapterWithFallback(_ bookId: BookID, _ chapter: Int) async throws -> [BiblePassage] {
            do {
                return try await remote.chapter(bookId: bookId, chapter: chapter)
            } catch BibleClientError.networkUnavailable {
                guard let offline else { throw BibleClientError.networkUnavailable }
                return try await offline.chapter(bookId: bookId, chapter: chapter)
            }
        }

        return BibleClient(
            passage: { reference in
                do {
                    return try await remote.passage(reference)
                } catch BibleClientError.networkUnavailable {
                    guard let offline else { throw BibleClientError.networkUnavailable }
                    return try await offline.passage(reference: reference)
                }
            },
            chapter: { bookId, chapter in try await chapterWithFallback(bookId, chapter) }
        )
    }
}

File: android/core/clients/src/main/kotlin/com/nexussoft/verbum/clients/helloao/HelloAOBibleClient.kt
package com.nexussoft.verbum.clients.helloao

import com.nexussoft.verbum.clients.BibleClient
import com.nexussoft.verbum.clients.BibleClientException
import com.nexussoft.verbum.clients.BundledBibleClient
import com.nexussoft.verbum.models.BibleBook
import com.nexussoft.verbum.models.BiblePassage
import com.nexussoft.verbum.models.BookId
import com.nexussoft.verbum.models.BookLanguage
import com.nexussoft.verbum.models.PassageReference
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

/** Fetches the text at a URL; injected so tests never touch the network. */
fun interface Transport {
    suspend fun get(url: String): String
}

/** Plain `HttpURLConnection`: a GET of a static file is all this API needs. */
object UrlConnectionTransport : Transport {
    override suspend fun get(url: String): String = withContext(Dispatchers.IO) {
        val connection = URL(url).openConnection() as HttpURLConnection
        try {
            connection.connectTimeout = 10_000
            connection.readTimeout = 15_000
            if (connection.responseCode == 404) throw BibleClientException.ContentUnavailable(PassageReference(url.substringAfterLast('/'), 0))
            connection.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
        } finally {
            connection.disconnect()
        }
    }
}

/**
 * Reads Scripture from bible.helloao.org and keeps every chapter it has seen in [cache]
 * so it reads again offline (§39). Mirrors iOS `HelloAOBibleClient`.
 */
class HelloAOBibleClient(
    val translationId: String,
    private val transport: Transport = UrlConnectionTransport,
    private val cache: ChapterCache = ChapterCache(null),
) : BibleClient {
    companion object {
        const val BASE_URL = "https://bible.helloao.org/api"
    }

    override suspend fun chapter(bookId: BookId, chapter: Int): List<BiblePassage> = withContext(Dispatchers.IO) {
        val book = BibleBook.book(bookId) ?: throw BibleClientException.UnknownBook(bookId)
        val reference = PassageReference(bookId, chapter)
        val usfm = HelloAOBooks.usfmByOsis[bookId]
        if (chapter !in 1..book.chapterCount || usfm == null) throw BibleClientException.ContentUnavailable(reference)
        cache.read(translationId, reference)?.takeIf { it.isNotEmpty() && it.all { verse -> verse.translationId == translationId } }?.let { return@withContext it }
        val body = try {
            transport.get("$BASE_URL/$translationId/$usfm/$chapter.json")
        } catch (e: BibleClientException) {
            throw e
        } catch (e: kotlinx.coroutines.CancellationException) {
            throw e
        } catch (e: Exception) {
            throw BibleClientException.NetworkUnavailable
        }
        val passages = HelloAOChapter.passages(body, bookId)
        if (passages.isEmpty() || passages.any { it.translationId != translationId }) throw BibleClientException.ContentUnavailable(reference)
        cache.write(translationId, reference, passages)
        passages
    }

    override suspend fun passage(reference: PassageReference): BiblePassage {
        val verses = chapter(reference.bookId, reference.chapter)
        val range = reference.verses ?: 1..verses.size
        if (range.last > verses.size) throw BibleClientException.VerseOutOfRange(reference, verses.size)
        val rangeText = if (range.first == range.last) "${range.first}" else "${range.first}-${range.last}"
        return BiblePassage(
            "$translationId:${reference.bookId}.${reference.chapter}.$rangeText", translationId, reference.bookId, reference.chapter,
            range.first, range.last, verses.subList(range.first - 1, range.last).joinToString(" ") { it.text },
        )
    }
}

/** Chapters already read, as JSON files in [directory] (null = memory only). */
class ChapterCache(private val directory: File?) {
    private val memory = object : LinkedHashMap<String, List<BiblePassage>>(24, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, List<BiblePassage>>): Boolean = size > 24
    }
    private val json = Json
    private val serializer = ListSerializer(BiblePassageJson.serializer())

    @Synchronized
    fun read(translationId: String, reference: PassageReference): List<BiblePassage>? {
        val key = key(translationId, reference)
        memory[key]?.let { return it }
        val file = file(key)?.takeIf { it.exists() } ?: return null
        return runCatching { json.decodeFromString(serializer, file.readText()).map { it.toModel() } }.getOrNull()?.also { memory[key] = it }
    }

    @Synchronized
    fun write(translationId: String, reference: PassageReference, passages: List<BiblePassage>) {
        val key = key(translationId, reference)
        memory[key] = passages
        val file = file(key) ?: return
        runCatching {
            file.parentFile?.mkdirs()
            file.writeText(json.encodeToString(serializer, passages.map { BiblePassageJson.of(it) }))
        }
    }

    private fun key(translationId: String, reference: PassageReference) = "${translationId}_${reference.bookId}_${reference.chapter}"
    private fun file(key: String) = directory?.let { File(it, "$key.json") }
}

@kotlinx.serialization.Serializable
internal data class BiblePassageJson(val id: String, val translationId: String, val bookId: String, val chapter: Int, val verseStart: Int, val verseEnd: Int, val text: String) {
    fun toModel() = BiblePassage(id, translationId, bookId, chapter, verseStart, verseEnd, text)
    companion object {
        fun of(p: BiblePassage) = BiblePassageJson(p.id, p.translationId, p.bookId, p.chapter, p.verseStart, p.verseEnd, p.text)
    }
}

/**
 * Production client: helloao in the device language, with two safety nets — chapters read
 * before (cache) and, for English, the bundled WEB — so the reader never dead-ends offline.
 */
class LiveBibleClient(language: BookLanguage, private val remote: BibleClient = HelloAOBibleClient(HelloAOTranslation.id(language))) : BibleClient {
    private val offline: BibleClient? = if (language == BookLanguage.ENGLISH) BundledBibleClient else null

    override suspend fun chapter(bookId: BookId, chapter: Int): List<BiblePassage> = try {
        remote.chapter(bookId, chapter)
    } catch (e: BibleClientException.NetworkUnavailable) {
        offline?.chapter(bookId, chapter) ?: throw e
    }

    override suspend fun passage(reference: PassageReference): BiblePassage = try {
        remote.passage(reference)
    } catch (e: BibleClientException.NetworkUnavailable) {
        offline?.passage(reference) ?: throw e
    }
}
